python/comps.py

```diff
@@ -1,36 +1,83 @@
+    lightgbm leaf node assignments (`observation_df`) compared to a second
+    dataframe of leaf node assignments (`comparison_df`).
+
+    Leaf nodes are weighted according to a tree importance matrix `weights`
+    and used to generate a similarity score. The function returns two
+    dataframes: One containing the indices of the most similar compararables
+    and the other containing their corresponding similarity scores.
+
+    More details on the underlying algorithm can be found here:
+    https://ccao-data.github.io/lightsnip/articles/finding-comps.html
+
+    Args:
+        observation_df (pandas.DataFrame):
+            DataFrame containing leaf node assignments for observations.
+        comparison_df (pandas.DataFrame):
+            DataFrame containing leaf node assignments for potential
+            comparables.
+        weights (numpy.ndarray):
+            Importance weights for leaf nodes, used to compute similarity
+            scores.
+        num_comps (int, optional):
+            Number of top comparables to return for each observation.
+            Default is 5.
+        num_chunks (int, optional):
+            Number of chunks to split observations for progress reporting.
+            Default is 10.
+
+    Returns:
+        tuple:
+            - pd.DataFrame:
+                DataFrame containing the indices of the `num_comps`
+                most similar comparables in `comparison_df`. The order of
+                rows will match the order of rows in `observation_df`.
+            - pd.DataFrame:
+                DataFrame containing similarity scores for the `num_comps`
+                most similar comparables. The order of rows will match the
+                order of rows in `observation_df`.
     """
+    # Check to make sure the shape of the input matrices is correct
+    if observation_df.shape[1] != comparison_df.shape[1]:
+        raise ValueError(
+            "Number of columns in `observation_df` "
+            f"({observation_df.shape[1]}) "
+            f"must match `comparison_df` ({comparison_df.shape[1]})"
+        )
+    if comparison_df.shape != weights.shape:
+        raise ValueError(
+            f"`comparison_df.shape` {comparison_df.shape} must match "
+            f"`weights.shape` {weights.shape}"
+        )
+
     # Convert the weights to a numpy array so that we can take advantage of
     # numba acceleration later on
     weights_matrix = np.asarray(weights, dtype=np.float32)
 
     # Chunk the observations so that the script can periodically report progress
-    num_chunks = 10
     observation_df["chunk"] = pd.cut(
         observation_df.index, bins=num_chunks, labels=False
     )
 
     total_num_observations = len(observation_df)
     total_num_possible_comps = len(comparison_df)
     chunked_ids, chunked_scores = [], []
-    for chunk_num in set(observation_df["chunk"]):
+    for chunk_num in observation_df["chunk"].unique():
```

Comment:
A subtle potential bug here: `set()` is not guaranteed to preserve the order of observations, while `Series.unique()` is. The order of the chunks is important here because we want the output rows to be in the exact same order as the input rows so that the caller can join the comps to the input observations; this means we always want to iterate through chunks in the order they appear in the input data. This affected my chunking-related unit tests, but I don't think it has affected prod runs; the CloudWatch logs for those runs seem to have printed chunks in the order I would expect.